From 0ee1c1b1e0ef1fe6d7d4e7923a28a2a01fb5125a Mon Sep 17 00:00:00 2001
From: Chau Ly <chaul@amperecomputing.com>
Date: Wed, 10 Sep 2025 09:40:53 +0000
Subject: [PATCH 1/1] NVMeBasicContext: Parsing and applying SMBus PEC

Sometimes when the PCIe device is being linked up, it's not ready for
PCI Express Basic Management Command (SMBUS/I2C) but we still read into
the device file. This results in garbage data in the temperature sensor
value and can cause threshold events.

Following the NVM Express Management Interface Specification
document [1], this commit enables Packet-Error-Checking (PEC) in smbus
driver (An 8 bit CRC calculated over the SMBus address, command code,
second SMBus address, and returned data) for NVMe sensors to ensure the
integrity of data transmission. If this check fails, i2c smbus read to
the device file will return a negative error code and no garbage value
will be propagated to D-Bus.

This is supported as an Entity-Manager configuration as not all devices
have their firmwares support PEC. Always enabling this check can cause
constant PEC failure for PEC-unsupported devices. The property is named
`PEC` [2] and, if set to "Required", will mean that driver's SMBus PEC
generation and verification is enabled. If the property is "NotRequired"
or not configured, SMBus PEC will not be enabled.

[1]: https://nvmexpress.org/wp-content/uploads/NVMe_Management_-_Technical_Note_on_Basic_Management_Command.pdf
[2]: https://gerrit.openbmc.org/c/openbmc/entity-manager/+/85901

Tested on the PCIe NVMe device that supports PEC and sometimes reports
false threshold events during power cycle stress test

1. Configure the "PEC" property to "Required" for the NVMe sensor
    => sensor reading works normally with valid sensor data.
2. Stress test turning off and on the power to the NVMe device =>
   There's no threshold events randomly reported for the sensor.
3. Reconfigure the sensor with "PEC": "NotRequired" and re-run step 2 =>
   Sometimes, unexpected threshold events are reported for the sensor
   with unexpected data.

Change-Id: I52f5ad464d09f68da25f1f105032210545c5eb4b
Signed-off-by: Chanh Nguyen <chanh@os.amperecomputing.com>
Signed-off-by: Chau Ly <chaul@amperecomputing.com>
---
 src/nvme/NVMeBasicContext.cpp | 68 ++++++++++++++++++++---------------
 src/nvme/NVMeSensor.cpp       | 18 +++++-----
 src/nvme/NVMeSensor.hpp       |  3 +-
 src/nvme/NVMeSensorMain.cpp   | 21 ++++++++++-
 4 files changed, 72 insertions(+), 38 deletions(-)

diff --git a/src/nvme/NVMeBasicContext.cpp b/src/nvme/NVMeBasicContext.cpp
index 509407a..4b81b36 100644
--- a/src/nvme/NVMeBasicContext.cpp
+++ b/src/nvme/NVMeBasicContext.cpp
@@ -3,7 +3,6 @@
 #include "NVMeContext.hpp"
 #include "NVMeSensor.hpp"
 
-#include <endian.h>
 #include <sys/ioctl.h>
 #include <unistd.h>
 
@@ -48,37 +47,37 @@ extern "C"
  * https://nvmexpress.org/wp-content/uploads/NVMe_Management_-_Technical_Note_on_Basic_Management_Command.pdf
  */
 
-static std::shared_ptr<std::array<uint8_t, 6>> encodeBasicQuery(
-    int bus, uint8_t device, uint8_t offset)
+struct BasicQueryCmd
+{
+    int bus;
+    uint8_t device;
+    uint8_t offset;
+    bool smbusPEC;
+};
+
+static std::shared_ptr<BasicQueryCmd> encodeBasicQuery(
+    int bus, uint8_t device, uint8_t offset, bool smbusPEC)
 {
     if (bus < 0)
     {
         throw std::domain_error("Invalid bus argument");
     }
 
-    /* bus + address + command */
-    uint32_t busle = htole32(static_cast<uint32_t>(bus));
-    auto command =
-        std::make_shared<std::array<uint8_t, sizeof(busle) + 1 + 1>>();
-    memcpy(command->data(), &busle, sizeof(busle));
-    (*command)[sizeof(busle) + 0] = device;
-    (*command)[sizeof(busle) + 1] = offset;
-
+    BasicQueryCmd tmpCmd = {bus, device, offset, smbusPEC};
+    auto command = std::make_shared<BasicQueryCmd>(tmpCmd);
     return command;
 }
 
-static void decodeBasicQuery(const std::array<uint8_t, 6>& req, int& bus,
-                             uint8_t& device, uint8_t& offset)
+static void decodeBasicQuery(const BasicQueryCmd& req, int& bus,
+                             uint8_t& device, uint8_t& offset, bool& smbusPEC)
 {
-    uint32_t busle = 0;
-
-    memcpy(&busle, req.data(), sizeof(busle));
-    bus = le32toh(busle);
-    device = req[sizeof(busle) + 0];
-    offset = req[sizeof(busle) + 1];
+    bus = req.bus;
+    device = req.device;
+    offset = req.offset;
+    smbusPEC = req.smbusPEC;
 }
 
-static void execBasicQuery(int bus, uint8_t addr, uint8_t cmd,
+static void execBasicQuery(int bus, uint8_t addr, uint8_t cmd, bool smbusPEC,
                            std::vector<uint8_t>& resp)
 {
     int32_t size = 0;
@@ -99,6 +98,18 @@ static void execBasicQuery(int bus, uint8_t addr, uint8_t cmd,
             return;
         }
 
+        if (smbusPEC)
+        {
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
+            if (::ioctl(fileHandle.handle(), I2C_PEC, 1) < 0)
+            {
+                lg2::error(
+                    "Could not set PEC to device '{ADDR}' for bus '{BUS}': '{ERRNO}'",
+                    "ADDR", lg2::hex, addr, "BUS", bus, "ERRNO", errno);
+                return;
+            }
+        }
+
         resp.resize(UINT8_MAX + 1);
 
         /* Issue the NVMe MI basic command */
@@ -142,13 +153,13 @@ static ssize_t processBasicQueryStream(FileHandle& in, FileHandle& out)
         uint8_t offset = 0;
         uint8_t len = 0;
         int bus = 0;
+        bool smbusPEC = false;
 
-        /* bus + address + command */
-        std::array<uint8_t, sizeof(uint32_t) + 1 + 1> req{};
+        BasicQueryCmd req{};
 
         /* Read the command parameters */
-        ssize_t rc = ::read(in.handle(), req.data(), req.size());
-        if (rc != static_cast<ssize_t>(req.size()))
+        ssize_t rc = ::read(in.handle(), &req, sizeof(req));
+        if (rc != static_cast<ssize_t>(sizeof(req)))
         {
             lg2::error(
                 "Failed to read request from in descriptor '{ERROR_MESSAGE}'",
@@ -160,10 +171,10 @@ static ssize_t processBasicQueryStream(FileHandle& in, FileHandle& out)
             return -EIO;
         }
 
-        decodeBasicQuery(req, bus, device, offset);
+        decodeBasicQuery(req, bus, device, offset, smbusPEC);
 
         /* Execute the query */
-        execBasicQuery(bus, device, offset, resp);
+        execBasicQuery(bus, device, offset, smbusPEC, resp);
 
         /* Write out the response length */
         len = resp.size();
@@ -285,11 +296,12 @@ void NVMeBasicContext::readAndProcessNVMeSensor()
         return;
     }
 
-    auto command = encodeBasicQuery(sensor->bus, sensor->address, 0x00);
+    auto command =
+        encodeBasicQuery(sensor->bus, sensor->address, 0x00, sensor->smbusPEC);
 
     /* Issue the request */
     boost::asio::async_write(
-        reqStream, boost::asio::buffer(command->data(), command->size()),
+        reqStream, boost::asio::buffer(command.get(), sizeof(*command)),
         [command](boost::system::error_code ec, std::size_t) {
             if (ec)
             {
diff --git a/src/nvme/NVMeSensor.cpp b/src/nvme/NVMeSensor.cpp
index be12f90..92a7c63 100644
--- a/src/nvme/NVMeSensor.cpp
+++ b/src/nvme/NVMeSensor.cpp
@@ -36,17 +36,19 @@
 static constexpr double maxReading = 127;
 static constexpr double minReading = 0;
 
-NVMeSensor::NVMeSensor(sdbusplus::asio::object_server& objectServer,
-                       boost::asio::io_context& /*unused*/,
-                       std::shared_ptr<sdbusplus::asio::connection>& conn,
-                       const std::string& sensorName,
-                       std::vector<thresholds::Threshold>&& thresholdsIn,
-                       const std::string& sensorConfiguration,
-                       const int busNumber, const uint8_t slaveAddr) :
+NVMeSensor::NVMeSensor(
+    sdbusplus::asio::object_server& objectServer,
+    boost::asio::io_context& /*unused*/,
+    std::shared_ptr<sdbusplus::asio::connection>& conn,
+    const std::string& sensorName,
+    std::vector<thresholds::Threshold>&& thresholdsIn,
+    const std::string& sensorConfiguration, const int busNumber,
+    const uint8_t slaveAddr, bool smbusPEC) :
     Sensor(escapeName(sensorName), std::move(thresholdsIn), sensorConfiguration,
            NVMeSensor::sensorType, false, false, maxReading, minReading, conn,
            PowerState::on),
-    bus(busNumber), address(slaveAddr), objServer(objectServer)
+    bus(busNumber), address(slaveAddr), smbusPEC(smbusPEC),
+    objServer(objectServer)
 {
     if (bus < 0)
     {
diff --git a/src/nvme/NVMeSensor.hpp b/src/nvme/NVMeSensor.hpp
index 31afbf3..6427e99 100644
--- a/src/nvme/NVMeSensor.hpp
+++ b/src/nvme/NVMeSensor.hpp
@@ -23,7 +23,7 @@ class NVMeSensor : public Sensor
                const std::string& sensorName,
                std::vector<thresholds::Threshold>&& thresholds,
                const std::string& sensorConfiguration, int busNumber,
-               uint8_t slaveAddr);
+               uint8_t slaveAddr, bool smbusPEC);
     ~NVMeSensor() override;
 
     NVMeSensor& operator=(const NVMeSensor& other) = delete;
@@ -32,6 +32,7 @@ class NVMeSensor : public Sensor
 
     const int bus;
     const uint8_t address;
+    const bool smbusPEC;
 
   private:
     const unsigned int scanDelayTicks = 5 * 60;
diff --git a/src/nvme/NVMeSensorMain.cpp b/src/nvme/NVMeSensorMain.cpp
index 76944cb..2182a4d 100644
--- a/src/nvme/NVMeSensorMain.cpp
+++ b/src/nvme/NVMeSensorMain.cpp
@@ -100,6 +100,18 @@ static std::optional<std::string> extractSensorName(
     return std::get<std::string>(findSensorName->second);
 }
 
+static std::optional<std::string> extractPEC(
+    const SensorBaseConfigMap& properties)
+{
+    auto findPEC = properties.find("PEC");
+    if (findPEC == properties.end())
+    {
+        return std::nullopt;
+    }
+
+    return std::visit(VariantToStringVisitor(), findPEC->second);
+}
+
 static std::filesystem::path deriveRootBusPath(int busNumber)
 {
     return "/sys/bus/i2c/devices/i2c-" + std::to_string(busNumber) +
@@ -193,6 +205,13 @@ static void handleSensorConfigurations(
                        *sensorName);
         }
 
+        bool smbusPEC = false;
+        std::optional<std::string> smbusPECStr = extractPEC(sensorConfig);
+        if (smbusPECStr)
+        {
+            smbusPEC = (*smbusPECStr == "Required");
+        }
+
         try
         {
             // May throw for an invalid rootBus
@@ -205,7 +224,7 @@ static void handleSensorConfigurations(
                 std::make_shared<NVMeSensor>(
                     objectServer, io, dbusConnection, *sensorName,
                     std::move(sensorThresholds), interfacePath, *busNumber,
-                    slaveAddr);
+                    slaveAddr, smbusPEC);
 
             context->addSensor(sensorPtr);
         }
-- 
2.31.1

